walk/walk.go

```diff
@@ -109,213 +110,383 @@ type WalkFunc func(dir, rel string, c *config.Config, update bool, f *rule.File,
+	case UpdateSubdirsMode:
+		_, ok := w.shouldUpdateRel[rel]
+		return ok || updateParent
+	default: // UpdateDirsMode
+		_, ok := w.shouldUpdateRel[rel]
+		return ok
+	}
 }
 
-// shouldCall returns true if Walk should call the callback in the
-// directory rel.
-func (u *UpdateFilter) shouldCall(rel string, updateParent bool) bool {
-	switch u.mode {
+// shouldCall returns whether the caller's Walk2Func callback should be called
+// on rel. We always need to call it on directories requested by the caller.
+// We may need to call it on their subdirectories, depending on mode. We also
+// need to call it on any additional directories requested by the callback.
+func (w *walker) shouldCall(rel string, updateParent bool) bool {
+	if w.visits[rel].containedByParent {
+		return false
+	}
+	switch w.mode {
 	case VisitAllUpdateSubdirsMode, VisitAllUpdateDirsMode:
 		return true
 	case UpdateSubdirsMode:
-		return updateParent || u.updateRels[rel]
+		return updateParent || w.shouldUpdateRel[rel]
 	default: // UpdateDirsMode
-		return u.updateRels[rel]
+		return w.shouldUpdateRel[rel]
 	}
 }
 
 // shouldUpdate returns true if Walk should pass true to the callback's update
 // parameter in the directory rel. This indicates the build file should be
 // updated.
-func (u *UpdateFilter) shouldUpdate(rel string, updateParent bool) bool {
-	if (u.mode == VisitAllUpdateSubdirsMode || u.mode == UpdateSubdirsMode) && updateParent {
+func (w *walker) shouldUpdate(rel string, updateParent bool) bool {
+	if (w.mode == VisitAllUpdateSubdirsMode || w.mode == UpdateSubdirsMode) && updateParent {
 		return true
 	}
-	return u.updateRels[rel]
+	return w.shouldUpdateRel[rel]
 }
 
-// shouldVisit returns true if Walk should visit the subdirectory rel.
-func (u *UpdateFilter) shouldVisit(rel string, updateParent bool) bool {
-	switch u.mode {
-	case VisitAllUpdateSubdirsMode, VisitAllUpdateDirsMode:
-		return true
-	case UpdateSubdirsMode:
-		_, ok := u.updateRels[rel]
-		return ok || updateParent
-	default: // UpdateDirsMode
-		_, ok := u.updateRels[rel]
-		return ok
+// visit is the main recursive function of walker. It visits one directory,
+// possibly recurses into subdirectories, and possible calls the callback.
+//
+// updateParent should indicate whether the the current mode tells Gazelle
+// to call the callback in the parent directory with update = true (see
+// shouldUpdate). The callback may not actually be called if the build file
+// contains syntax errors or a gazelle:ignore directive.
+func (w *walker) visit(c *config.Config, rel string, updateParent bool) {
+	w.visits[rel] = visitInfo{c: c} // to be updated when we're further along.
+
+	// Absolute path to the directory being visited
+	dir := filepath.Join(c.RepoRoot, rel)
+
+	// Load the build file.
+	info, err := w.cache.get(rel, w.loadDirInfo)
+	if err != nil {
+		w.errs = append(w.errs, err)
+	}
+	hasBuildFileError := err != nil
+
+	configure(w.cexts, w.knownDirectives, c, rel, info.file, info.config)
```

Comment:
I believe previously [`configure` was not called `if containedByParent`](https://github.com/bazel-contrib/bazel-gazelle/blob/a39914bf8d3d1b953ef79661df1b787b677b209a/walk/walk.go#L159-L162)

Reply:
Fixed